cloudevents/sdk-python
# Update types and handle data_base64 structured.

- Add sane defaults for encoding
  - Unfortunately, defaults for structured and binary need to be *different*
   - Push types through interfaces
- Make it easy to call 'ToRequest' using Marshaller defaults
- Add tests for above

- Resolves #28  and slighly improves #26 

- I discovered that the types for `data_marshaller` and `data_unmarshaller` had to vary between structured and unstructured as a *user* of the SDK. In particular, it seemed that `UnmarshalJSON` wants to pass a `str` value to `data_unmarshaller`, while `UnmarshalBinary` wants to pass a `typing.IO`.

- I started by figuring out the necessary types for `data_marshaller` and `data_unmarshaller`. Along the way, I realized that binary CloudEvents data was not being handled correctly for JSON. See #28 and https://github.com/cloudevents/spec/blob/v1.0/json-format.md#31-handling-of-data. Annoyingly, if the data is not of type "binary", the structured format suggests passing it through without an extra layer of enveloping, which leads to some special cases in `MarshalJSON`.

- How to verify it

I added tests

- One line description for the changelog

```
Update python types and handle structured `data_base64`.
```


Note that this is probably **a breaking API change** because the expected types for `data_marshaller` and `data_unmarshaller` changed and all previous clients needed to supply these.
